src/weighted_list.h

```diff
@@ -0,0 +1,138 @@
+#ifndef WEIGHTED_LIST_H
+#define WEIGHTED_LIST_H
+
+#include "rng.h"
+#include <vector>
+#include <functional>
+
+template <typename T> struct weighted_object {
+    weighted_object(const T &obj, const int weight) {
```

Comment:
Please use a member initializer list. The current code calls the default constructor of `T` and later copy assigns the actual value. This can be combined into one step: `weighted_object(...) : obj( obj ), ...`

Beside that, it looks fine. The current implementation with a simple int as weight should be good as first step, the lists with special weights can be added later.